### PR TITLE
feat(frontend): Add UserProtected Route

### PR DESCRIPTION
**Pull request by Pratik2026**
## Overview

1. This PR fixes or fixes part of #49 .
2. This PR does the following: It adds UserProtectedRoute to a demoPage(open when user clicks on 'Pay now') which redirect the user to a homePage as well as  pops a toaster if the user is not logged in otherwise opens the wrapped component .

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] The PR is made from a branch that's **not** called "main/master".

## Proof that changes are correct
[ProtectedRoutes.webm](https://github.com/bsoc-bitbyte/busify/assets/128511266/33e3e63e-4908-4556-a33a-ef1e80e25345)

### CODE REVIEW

File: frontend/src/App.tsx [hunk 1]
```diff
@@ -21,6 +25,14 @@ function App() {
           <Route path="/google" element={<GoogleAuthLogin />} />
           <Route path="/bus-schedule" element={<BusSchedule />} />
           <Route path="/bus-details" element={<BusDetails />} />
+          <Route
+            path="/demo-Page"
+            element={
+              <UserProtectedRoute isAuth={isAuth}>
```
Comment (prajjwalkapoor) [inline frontend/src/App.tsx:31]: Don't pass `isAuth` as prop. Write this logic inside the `UserProtectedRoute` component to make it independent.

**Review by prajjwalkapoor — APPROVED** [review]

**Review by prajjwalkapoor — CHANGES_REQUESTED** [review]
Also, if we directly go into the demo page by URL, it redirects us to the home page even when we are logged in. Look into it and see how we can fix this.

**Comment by Pratik2026** [comment]
@prajjwalkapoor It seems that when the user directly access the demo page, the user gets redirected to the home page as per the default value of isAuth state which is 'false'. This is due to the time delay in getting a response from checkAuth function. One possible solution would be to introduce a loading state which keeps the loading true until the response has been received from the server and then render the component accordingly.

**Comment by prajjwalkapoor** [comment]
> @prajjwalkapoor It seems that when the user directly access the demo page, the user gets redirected to the home page as per the default value of isAuth state which is 'false'. This is due to the time delay in getting a response from checkAuth function. One possible solution would be to introduce a loading state which keeps the loading true until the response has been received from the server and then render the component accordingly. Let me know what you think about it.

yeah, we can do this.

File: frontend/src/App.tsx (hunk 1 above)
Comment (Pratik2026) [reply frontend/src/App.tsx]: Done

**Pull request merged**